### PR TITLE
Add `AdamW` to C++ frontend

### PR DESCRIPTION
**Pull request by sotlampr**
Slightly modified Adam, following the python implementation, and the `ProducesPyTorchValues` tests pass. I had a problem with another test though (see commit c1a6241676ab84fc531c1c3a10f964aa5704092e), it seems that optimizing for two steps with the same optimizer vs optimizing for two steps using freshly initialized objects will produce the same output.

### CODE REVIEW

**Comment by glaringlee** [comment]
@sotlampr 
It seems this PR also included other people's changes. Can you rebase your branch to include your change only? i will review this ASAP. thx

**Comment by sotlampr** [comment]
Yeah, I rebased to pass some broken tests. Should be good now. Thanks!

File: test/cpp/api/serialize.cpp [hunk 1]
```diff
@@ -600,6 +600,56 @@ TEST(SerializeTest, Optim_Adam) {
   is_optimizer_state_equal<AdamParamState>(optim1.state(), optim1_2.state());
 }
 
+TEST(SerializeTest, Optim_AdamW) {
+  // test_serialize_optimizer<AdamW, AdamWOptions, AdamWParamState>(AdamWOptions().lr(0.99999).amsgrad(true).weight_decay(0.5));
```
Comment (glaringlee) [inline test/cpp/api/serialize.cpp:604]: This might be a precision issue, probably the testing parameter is not good enough to identify the difference. Can you try different parameters?

**Review by glaringlee — CHANGES_REQUESTED** [review]

File: test/cpp/api/optim_baseline.py [hunk 2]
```diff
@@ -26,6 +26,9 @@
     "Adam": lambda p: torch.optim.Adam(p, 1.0),
     "Adam_with_weight_decay": lambda p: torch.optim.Adam(p, 1.0, weight_decay=1e-2),
     "Adam_with_weight_decay_and_amsgrad": lambda p: torch.optim.Adam(p, 1.0, weight_decay=1e-6, amsgrad=True),
+    "AdamW": lambda p: torch.optim.AdamW(p, 1.0),
+    "AdamW_without_weight_decay": lambda p: torch.optim.AdamW(p, 1.0, weight_decay=0),
```
Comment (glaringlee) [inline test/cpp/api/optim_baseline.py:30]: Can we have a test case with non-zero decay?

**Review by glaringlee — CHANGES_REQUESTED** [review]

File: test/cpp/api/serialize.cpp (hunk 1 above)
Comment (sotlampr) [reply test/cpp/api/serialize.cpp:604]: Yes, and that is the case for e.g. (original) Adam with no weight decay, but it seems the right hyper-parameters are used in order for the test to pass (i.e. in `test/cpp/api/serialize.cpp:555` `weight_decay` is set to `0.5`). I ran a grid search and found the parameters that will make actually AdamW produce different outputs (hint: low `beta2`). It seems we have two options:

- Use the ad-hoc hyper-parameters that will produce different outpus
- Modify all tests to use the default hyper-parameters and modify the `test_serialize_optimizer` api to check for *stateful optimizers* or something simillar.

For now, I modified the test hyper-parameters in order to pass. If we want changes in `test_serialize_optimizer`, let's open another issue and I can work on it!

File: test/cpp/api/optim_baseline.py (hunk 2 above)
Comment (sotlampr) [reply test/cpp/api/optim_baseline.py:30]: The default `weight_decay` is set to `1e-2` in the python implementation, and consequently in mine. Should I write a test with custom weight decay `AdamW_with_custom_weight_decay` (naming it `AdamW_with_weight_decay` will be a bit confusing IMO since the default `AdamW` already has a non-zero weight decay)?

Comment (glaringlee) [reply test/cpp/api/optim_baseline.py:30]: Ah, I see what you mean, no need to change here then. The default case already have a non-zero weight_decay being tested. Thanks.

File: test/cpp/api/serialize.cpp (hunk 1 above)
Comment (glaringlee) [reply test/cpp/api/serialize.cpp]: I think for now, we can use ad-hoc hyper-parameters. Your change looks good.
The test_serialize_optimizer need some tweaks, it should check predictable state only. 
I will check how we want to refactor it and file issue later if necessary.

**Review by glaringlee — APPROVED** [review]
This is good to go now.
@sotlampr 
Thank you so much for doing this work !

File: test/cpp/api/serialize.cpp (hunk 1 above)
Comment (sotlampr) [reply test/cpp/api/serialize.cpp]: Cool - if you figure it out and need a hand with that drop a mention!

**Comment by sotlampr** [comment]
Awesome, thanks @glaringlee
